### DISCUSSION TITLE
Inference and embeddings directly on web browser with wasm

### DISCUSSION THREAD

**Discussion by ngxson** [ggml-org/llama.cpp#6055] [Show and tell]
Hi everyone,

I've just made a WebAssembly binding for llama.cpp that can run completely on browser. It already have support for high-level API like completions, embeddings, as well as some low-level functions like sampling, (de)tokenization,...

Wrapper code is written in typescript, can be embedded inside a ReactJS project (there's an example on my repo).

Would be nice if someone can play around and make a complete RAG-in-browser demo :-)

Link to the project: https://github.com/ngxson/wllama

**Comment by phymbert** [comment]
Wow, nice job. I will give it a try. I see in the [TODO](https://github.com/ngxson/wllama?tab=readme-ov-file#todo): `Guide: How to split gguf file?`
We just discussed it in:
 - https://github.com/ggerganov/llama.cpp/discussions/6000

Are you working on something similar ?

**Reply by ngxson** [reply]
I did read your proposal this morning, thanks for bringing it up. Yeah it would be nice to split gguf and still loadable by llama.cpp without doing `cat`

The "split" feature that I mentioned in my repo is simply `split` & `cat`. I added it mostly for networking performance, since splitting file is a common technique when talking about web.

**Comment by phymbert** [comment]
FYI: https://huggingface.co/docs/hub/gguf

**Reply by ngxson** [reply]
Yeah thanks for the info. In fact, I work on gguf format when I was making the `merge` tool #5741 so I know more or less the file layout.

The tricky part is that tensor info holds the offset of tensor inside a file, so if we split the file into multiple parts, we cannot reuse the same metadata table (since offset is changed).

In https://github.com/ggerganov/llama.cpp/discussions/6000 I proposed to have separated gguf context (that means each gguf has its own table of tensor info), so the offset will be correct. At high level it should be a simple change, but I've got time to dive deep into that. If you're not actively working on this, please let me know so maybe I'll have a look this weekend or next week.

**Reply by phymbert** [reply]
II will ping you on the PR for support
